### PR TITLE
[SPARK-34891][SS] Introduce state store manager for session window in streaming query

### PR DESCRIPTION
**Pull request by HeartSaVioR**
Introduction: this PR is a part of SPARK-10816 (`EventTime based sessionization (session window)`). Please refer #31937 to see the overall view of the code change. (Note that code diff could be diverged a bit.)

### What changes were proposed in this pull request?

This PR introduces state store manager for session window in streaming query. Session window in batch query wouldn't need to leverage state store manager.

This PR ensures versioning on state format for state store manager, so that we can apply further optimization after releasing Spark version. StreamingSessionWindowStateManager is a trait defining the available methods in session window state store manager. Its subclasses are classes implementing the trait with versioning.

The format of version 1 leverages the new feature of "prefix match scan" to represent the session windows:

* full key : [ group keys, start time in session window ]
* prefix key [ group keys ]

### Why are the changes needed?

This part is a one of required on implementing SPARK-10816.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New test suite added

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Except the test suite, one more thing worths to address here is write amplification; we "blindly" replace all start times and all sessions. This could bring unnecessary writes on "unmodified" existing sessions. In many cases we expect the new inputs will be bound and expanding to the existing sessions, but with very long watermark gap and old inputs which have various timestamps, the case could still happen.

**Comment by HeartSaVioR** [comment]
I just simply remedied write amplification a bit via determining existing values to be "deleted" or "replaced", instead of blindly judging it as "deleted".

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by viirya** [comment]
> I just simply remedied write amplification a bit via determining existing values to be "deleted" or "replaced", instead of blindly judging it as "deleted".

Does this reflect in this PR or original one?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+   * Remove using a predicate on values.
+   *
+   * At a high level, this produces an iterator over the values such that value satisfies the
+   * predicate, where producing an element removes the value from the state store and producing
+   * all elements with a given key updates it accordingly.
+   *
+   * This implies the iterator must be consumed fully without any other operations on this manager
+   * or the underlying store being interleaved.
+   */
+  def removeByValueCondition(removalCondition: UnsafeRow => Boolean): Iterator[UnsafeRow]
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:113]: Seems this is not in my implementation. I also don't find it in #31937. Am I missing it?

**Review by viirya — COMMENTED** [review]
I don't have too many questions for this. Most methods are familiar to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:113]: I moved the code on evicting state rows from store physical exec to state store impl, to separate the concern. Streaming join state store also does the same.

I think store exec shouldn't know about the details for state store itself, that said, this may need to be applied to store new state rows as well which is also in store exec. 
(I don't remember I addressed it in this PR. I'm sorry I'm not on my laptop and only mobile for following 3~4 days.)

**Comment by HeartSaVioR** [comment]
I addressed it in this PR. Please look for the recent commits to narrow the diff.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:113]: And this change is in #31937.

**Review by viirya — COMMENTED** [review]
Okay. By the understand of original context, this looks good to me. Although I'm afraid that for possible reviewers who aren't familiar with it, this might not be easy to digest.

cc @xuanyuanking maybe you can also look through this if you have some time.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
gentle ping @xuanyuanking

**Comment by xuanyuanking** [comment]
@viirya Sorry for the late. Just back from the work on open-sourcing RocksDBStateStore. Since the design has been published. I can back to this now.
I'm +1 for the `StreamingSessionWindowStateManager` abstraction. Only have some concern about the StreamingSessionWindowStateManagerImplV1. It would be great if we can have a better state store API to reduce the complexity of the implementation. I'll start working on it in these two weeks (in parallel with open-sourcing the RocksDB state store).

**Comment by viirya** [comment]
I think this is the last open sub-PRs for #31937. @HeartSaVioR @xuanyuanking so will this be changed after #33038 is merged?

**Comment by HeartSaVioR** [comment]
> I think this is the last open sub-PRs for #31937.

Unfortunately no. We have 4 open sub-tasks in [SPARK-10816](https://issues.apache.org/jira/browse/SPARK-10816) including newly added prefix scan, and I plan to submit a PR per sub-task. So in overall I still need to submit two more PRs.

**Comment by HeartSaVioR** [comment]
> so will this be changed after #33038 is merged?

Yes, I'll update this on top of #33038.

**Comment by viirya** [comment]
> We have 4 open sub-tasks in SPARK-10816

Do you mean open PRs or JIRAs? Based on #31937, I only see three open PRs and two PRs were merged.

**Comment by HeartSaVioR** [comment]
I haven't submitted two PRs based on this change. We will have more PRs soon.

**Comment by HeartSaVioR** [comment]
UPDATE: This is now on top of SPARK-35861 (#33038). I'll rebase this once we merge #33038.

**Comment by HeartSaVioR** [comment]
Changed this to draft for now to prevent merging before merging #33038

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
cc. @viirya @xuanyuanking I've rebased this one and this is ready to review. Please take a look. Thanks!

**Comment by viirya** [comment]
Thanks @HeartSaVioR. I'll review this today/tomorrow.

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+  /**
+   * Returns all sessions for the key.
+   *
+   * @param key The key without session, which can be retrieved from
+   *            {@code extractKeyWithoutSession}.
+   */
+  def getSessions(store: ReadStateStore, key: UnsafeRow): Iterator[UnsafeRow]
+
+  /**
+   * Replaces all sessions for the key to given one.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:67]: What the returned long is?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+  private val sessionProjection: UnsafeProjection =
+    GenerateUnsafeProjection.generate(Seq(sessionExpression), inputSchema)
+
+  // extract session_window (start, end) from UnsafeRow
+  def extractTimePair(value: InternalRow): (Long, Long) = {
+    val window = sessionProjection(value).getStruct(0, 2)
+    (window.getLong(0), window.getLong(1))
+  }
+
+  def removeOverlapWindow(
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:295]: Could you add some comment on this method?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+  private def getKey(value: UnsafeRow): UnsafeRow = keyRowGenerator(value)
+
+  private def getStateKey(key: UnsafeRow, sessionStart: Long): UnsafeRow = {
+    val stateKey = stateKeyRowGenerator(key)
+    stateKey.setLong(indexOrdinalInSessionStart, sessionStart)
+    stateKey.copy()
+  }
+
+  /**
+   * add to store, only update the changed window elements
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:226]: Could you rephrase the method doc? Maybe "Updates windows from the values overlapped with existing values in the store and adds new windows from the values."?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManagerSuite.scala [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  private val sessionAttribute = rowAttributes.filter(_.name == "session").head
+
+  private val inputValueGen = UnsafeProjection.create(rowAttributes.map(_.dataType).toArray)
+  private val inputKeyGen = UnsafeProjection.create(
+    keysWithoutSessionAttributes.map(_.dataType).toArray)
+
+  before {
+    SparkSession.setActiveSession(spark) // set this before force initializing 'joinExec'
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManagerSuite.scala:50]: `joinExec`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+   * Extracts the key without session window from the row.
+   * This can be used to group session windows by key.
+   */
+  def extractKeyWithoutSession(value: UnsafeRow): UnsafeRow
+
+  /**
+   * Check whether the given value is a new session, or the session has been
+   * updated.
+   * This can be used to control the output in UPDATE mode.
+   */
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:55]: This looks confusing. Maybe "Returns true if the session of the given value doesn't exist in the store, or the value in the session is different to the stored value of the session in the store."

**Review by viirya — COMMENTED** [review]
Some minor comments. After addressing them, this should be good to go. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+  @transient private lazy val keyWithoutSessionRowGenerator = UnsafeProjection.create(
+    keyWithoutSessionExpressions, valueAttributes)
+
+  private val stateKeyStructType = keyWithoutSessionExpressions.toStructType
+    .add("sessionStartTime", TimestampType, nullable = false)
+
+  private val stateKeyExprs = keyWithoutSessionExpressions :+ Literal(1L)
+  private val indexOrdinalInSessionStart = keyWithoutSessionExpressions.size
+  @transient private lazy val keyRowGenerator = UnsafeProjection.create(
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:139]: Nice finding. Looks like I missed to remove one later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:55]: Proposed sentence looks nice. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala:67]: The return value is just same as sessions.length so not that useful.

I think we'd want to measure the number of rows on "added/updated" vs "deleted" and we have both cases on executing method, which requires another sort of return type. Let's keep this as Unit and change it in session window implementation PR.

**Comment by HeartSaVioR** [comment]
Thanks @viirya for reviewing! I reflected review comments.

Btw, I also made a change via f12cda8 as I realized we don't need to union windows from both (old/new) sides.
(Actually I found this during experiment but I forgot to reflect here. My bad.)

Simply thinking, given old windows and new windows, we can simply remove all old windows and add all new windows as new windows are the thing we want to "retain" in state store. (See method contract on `updateSessions`.) Adding some optimization on reducing "delta"/"tombstone", we can remove old windows which state keys are not updated by new windows, and add new windows to do "upsert".

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks @HeartSaVioR. Merging to master/3.2.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging! I'll update the next PR. We have two PRs now.
